advice_animal/runner.py

```diff
@@ -11,10 +15,162 @@
 LOG = logging.getLogger(__name__)
 
 
+def run_cmd(cmd: list[Union[str, Path]], check: bool = True) -> Tuple[str, int]:
+    LOG.info("Run %s in %s", cmd, os.getcwd())
+    proc = subprocess.run(cmd, encoding="utf-8", capture_output=True, check=check)
+    LOG.debug("Ran %s -> %s", cmd, proc.returncode)
+    LOG.debug("Stdout: %s", proc.stdout)
+    return proc.stdout, proc.returncode
+
+
+@dataclass
+class Result:
+    advice_name: str
+    success: bool
+    project: str = ""
+    message: str = ""
+    changes_needed: bool = False
+
+
 class Runner:
-    def __init__(self, env: Env, check_dir: Path) -> None:
-        self.env = env
-        self.check_dir = check_dir
+    def __init__(
+        self, advice_path: Path, inplace: bool, mode: Literal["check", "diff", "apply"]
+    ) -> None:
+        if not advice_path.is_dir():
+            raise ValueError(f"{advice_path} is not a directory")
+
+        if inplace and mode != "apply":
+            raise ValueError("inplace only valid with mode 'apply'")
+
+        self.advice_path = advice_path
+        self.inplace = inplace
+        self.mode = mode
+
+    def run(
+        self,
+        repo: Path,
+        confidence_filter: FixConfidence,
+        preview_filter: bool,
+        name_filter: re.Pattern[str],
+    ):
+        """
+        1. Get a list of eligible checks
+        2. Clone the repo to a tempdir
+        3. For each check:
+            if inplace:
+                - Iterate over the python projects
+                - Run the check
+            else:
+                - Create a branch
+                - Iterate over the python projects
+                    - Run the check
+                    - git addremove
+                if mode == "diff":
+                    - git diff
+                elif mode == "apply":
+                    - git commit
+                    - git push
+                elif mode == "check":
+                    - git diff --exit-code
+
+        """
+        git_head_path = repo / ".git" / "HEAD"
+        if git_head_path.exists():
+            current_branch = git_head_path.read_text().strip().split("/")[-1]
+        else:  # This is a detached HEAD or not a git repo
+            current_branch = None
+            if not self.inplace:
+                raise ValueError("Not a git repo")
+        results = {}
+        if self.inplace:  # This is only applicable for `apply` command
+            cur_cwd = os.getcwd()
+            try:
+                for advice_name, check_cls in self.iter_check_classes(
+                    confidence_filter, preview_filter, name_filter
+                ):
+                    LOG.log(VLOG_1, "Running check %s", advice_name)
+                    env = Env(repo)
```

Comment:
I think Env should be renamed to ProjectEnv and moved inside the loop below.  It has at least the top-level python name (which will vary per project) and the .get cache someday, which should also be invalidated between each project.

Also in the non-inplace case.